### ISSUE TITLE
clang-analyzer: False positive when variable is cast multiple times (IS_ERR_VALUE() )

### ISSUE THREAD

**Issue by daniel-thompson** [llvm/llvm-project#55367]
TL;DR analyzer fails to reason that e is non-zero when the following predicate is true: e >= (unsigned long) -4095 . Normally thje analyzer can apply this reasoning but appears to be confused by the casting involved in the Linux kernel functions and macros that do this.

After running clang-analyzer on Linux kernel code I discovered a false positive where the analyzer simultaneously assumes the same value as both non-zero and zero in different parts of the reasoning tree. The false positive triggers on a common code pattern inside the Linux kernel (returning an -ve errno instead of a pointer). I have provided multiple versions of the failed reasoning all of which are effectively different implementations of the IS_ERR_VALUE() macro found in the Linux kernel. If I implement the checks as a signed long then the analyzer can reason correctly, however it I implement similar checks using an unsigned long then there is a false error report.

It appears that the analyzer would normally reason correctly (or at least not report a warning), however the frequent casting between pointer, long, unsigned long and int types is confusing the analyzer.

The following minimized example has been tested on both clang-13 and clang-14:

~~~ c
// Use these to control which implementation of IS_ERR() to use
//#define WANT_SIGNED_LONG_IS_ERR_VALUE
#define WANT_UNSIGNED_LONG_IS_ERR_VALUE

#define MAX_ERRNO 4095

static inline int IS_ERR(const void *ptr)
{
#if defined WANT_SIGNED_LONG_IS_ERR_VALUE

	/*
	 * This is semantically equivalent to the real IS_ERR_VALUE()
	 * implementation but works using a long rather than an unsigned long.
	 *
	 * This implementation does not provoke any false-positive.
	 */
	long e = (long) ptr;
	return e < 0 && e >= -MAX_ERRNO;

#elif defined WANT_UNSIGNED_LONG_IS_ERR_VALUE

	/*
	 * This is partway between the two IS_ERR_VALUE() implementations above
	 * and below and show the failed reasoning in the most obvious way.
	 *
	 * It is the same check as the one used in Linux kernel but prefixed with
	 * an explicit comparison to zero. Thus the output from the analyzer is
	 * easier to read since we can see it explicitly assume e in non-zero
	 * rather than implicitly by checking ptr is (unsigned) >= -4095.
	 *
	 * It implies that the failure to reason about the range of x is
	 * linked to casting it as an unsigned long.
	 */
	unsigned long e = (unsigned long) ptr;
	return e != 0 && e >= (unsigned long) -MAX_ERRNO;

#else

	/*
	 * This is the actual check that appears in the Linux kernel.
	 *
	 * By treating the value as unsigned then the range check can be a
	 * single comparison. Currently clang-analyzer does not use use this
	 * predicate to reason about the possible values of x.
	 *
	 * In particular clang-analyzer will explore code paths where
	 * IS_ERR_VALUE(x) is true but then later assume that x is zero.
	 */
	return ((unsigned long) (ptr) >= (unsigned long)-MAX_ERRNO);

#endif
}

#define EINVAL 22

struct task_struct {
	unsigned int ctrl;
};

static struct task_struct *ptrace_hbp_get_event(unsigned int note_type,
					       struct task_struct *tsk)
{
	return note_type ? tsk : (void *) -EINVAL;
}

static int ptrace_hbp_get_ctrl(unsigned int note_type,
			       struct task_struct *tsk,
			       unsigned int *ctrl)
{
	struct task_struct *t = ptrace_hbp_get_event(note_type, tsk);

	if (IS_ERR(t))
		// Note: the false positive can also be dismissed if we
		// remove the (long) here!
		return (int) (long) t;

	*ctrl = t->ctrl;
	return 0;
}

int hw_break_get(struct task_struct *target, unsigned int note_type,
		 unsigned int *to)
{
	int ret;
	unsigned int ctrl;

	ret = ptrace_hbp_get_ctrl(note_type, target, &ctrl);
	if (ret)
		return ret;
	*to = ctrl;
	return 0;
}
~~~

The test case output is easiest to read when WANT_UNSIGNED_LONG_IS_ERR_VALUE is set as shown below (with a couple of `<<<< chevrons >>>>` to highlight the key parts.

~~~
maple$ clang-14 -Wall --analyze --analyzer-output text -c ptrace-IS_ERR_VALUE.c 
ptrace-IS_ERR_VALUE.c:109:6: warning: Assigned value is garbage or undefined [core.uninitialized.Assign]
        *to = ctrl;
            ^ ~~~~
ptrace-IS_ERR_VALUE.c:104:2: note: 'ctrl' declared without an initial value
        unsigned int ctrl;
        ^~~~~~~~~~~~~~~~~
ptrace-IS_ERR_VALUE.c:106:8: note: Calling 'ptrace_hbp_get_ctrl'
        ret = ptrace_hbp_get_ctrl(note_type, target, &ctrl);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ptrace-IS_ERR_VALUE.c:89:26: note: Calling 'ptrace_hbp_get_event'
        struct task_struct *t = ptrace_hbp_get_event(note_type, tsk);
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ptrace-IS_ERR_VALUE.c:82:9: note: Assuming 'note_type' is not equal to 0
        return note_type ? tsk : (void *) -EINVAL;
               ^~~~~~~~~
ptrace-IS_ERR_VALUE.c:82:9: note: '?' condition is true
ptrace-IS_ERR_VALUE.c:82:2: note: Returning pointer, which participates in a condition later
        return note_type ? tsk : (void *) -EINVAL;
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ptrace-IS_ERR_VALUE.c:89:26: note: Returning from 'ptrace_hbp_get_event'
        struct task_struct *t = ptrace_hbp_get_event(note_type, tsk);
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ptrace-IS_ERR_VALUE.c:91:6: note: Calling 'IS_ERR'
        if (IS_ERR(t))
            ^~~~~~~~~
ptrace-IS_ERR_VALUE.c:54:9: note: Assuming 'e' is not equal to 0, which participates in a condition later
        return e != 0 && e >= (unsigned long) -MAX_ERRNO;           <<<<< e/ret is assumed non-zero >>>>>
               ^~~~~~
ptrace-IS_ERR_VALUE.c:54:9: note: Left side of '&&' is true
ptrace-IS_ERR_VALUE.c:54:19: note: Assuming the condition is true
        return e != 0 && e >= (unsigned long) -MAX_ERRNO;
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ptrace-IS_ERR_VALUE.c:54:2: note: Returning the value 1, which participates in a condition later
        return e != 0 && e >= (unsigned long) -MAX_ERRNO;
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ptrace-IS_ERR_VALUE.c:91:6: note: Returning from 'IS_ERR'
        if (IS_ERR(t))
            ^~~~~~~~~
ptrace-IS_ERR_VALUE.c:91:2: note: Taking true branch
        if (IS_ERR(t))
        ^
ptrace-IS_ERR_VALUE.c:94:3: note: Returning without writing to '*ctrl'
                return (int) (long) t;
                ^
ptrace-IS_ERR_VALUE.c:94:3: note: Returning value (loaded from 't'), which participates in a condition later
                return (int) (long) t;
                ^~~~~~~~~~~~~~~~~~~~~
ptrace-IS_ERR_VALUE.c:106:8: note: Returning from 'ptrace_hbp_get_ctrl'
        ret = ptrace_hbp_get_ctrl(note_type, target, &ctrl);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ptrace-IS_ERR_VALUE.c:107:6: note: Assuming 'ret' is 0
        if (ret)                                               <<<<< e/ret is later assumed to be zero >>>>>
            ^~~
ptrace-IS_ERR_VALUE.c:107:2: note: Taking false branch
        if (ret)
        ^
ptrace-IS_ERR_VALUE.c:109:6: note: Assigned value is garbage or undefined
        *to = ctrl;
            ^ ~~~~
1 warning generated.
~~~
